moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp

```diff
@@ -86,7 +86,8 @@ bool GlobalPlannerComponent::initializeGlobalPlanner()
         return rclcpp_action::CancelResponse::ACCEPT;
       },
       [this](std::shared_ptr<rclcpp_action::ServerGoalHandle<moveit_msgs::action::GlobalPlanner>> goal_handle) {
-        return globalPlanningRequestCallback(goal_handle);
+        // this needs to return quickly to avoid blocking the executor, so spin up a new thread
+        std::thread{ &GlobalPlannerComponent::globalPlanningRequestCallback, this, goal_handle }.detach();
```

Comment:
I wouldn't copy that from ros2... that seems about as bad as their spin_until_future_complete implementation.

https://stackoverflow.com/a/58222149

Doing this is technically creating a memory violation because the thread object might get cleaned up after this scope exits and its state is no longer valid.

Until we get `jthread` from c++20 which can detach a thread without creating a memory violation, we should instead always maintain the lifetime of thread object for as long as it runs and join it before cleaning up the memory.

Reply:
any suggestion on how to actually do that? I could see something like storing a vector of `std::threads`, and joining all of them in the destructor. Sigh

Reply:
OK, I added the std::thread as class member so there's no concern about scope. I also added a simple 30-line example of action server cancellation to have more confidence that it works. LGTM at this point.

Reply:
Thank you! Two small things before merging this. You can drop the unique_ptr here. Simply make the thread a member, the default constructor will leave it uninitialized. That saves you the nullptr logic. Instead, before joining, threads should be checked if they are joinable, otherwise it allows for an error condition https://en.cppreference.com/w/cpp/thread/thread/join